JAVA_OPTIONS with spaces (even with double quotes) fails with CNFE when running (for the second time)

Based on https://rmannibucau.metawerx.net/post/jul-production-ready-pattern-with-simpleformatter, I have a class with the following JAVA_OPTIONS set:
```java
//JAVA_OPTIONS "-Djava.util.logging.SimpleFormatter.format=%1$s [%4$s] %5$s%6$s%n"
```
When I run it the second time (without changing anything in the source), JBang fails with a CNFE
```java
~/workspace/quarkus-extension-catalog (main ✔) ᐅ jbang .github/workflows/publishcatalog.java --working-directory=. --registry-url=http://localhost:8080 --token=foo -d
[jbang] Building jar...
2021-03-31T10:14:42.703449-03:00[America/Sao_Paulo][WARN] Running in dry-run mode. No files will be changed or posted to the registry
2021-03-31T10:14:42.859159-03:00[America/Sao_Paulo][INFO] Processing platform ./platforms/quarkus-bom.yaml

~/workspace/quarkus-extension-catalog (main ✘)✹ ᐅ jbang .github/workflows/publishcatalog.java --working-directory=. --registry-url=http://localhost:8080 --token=foo -d
Error: Could not find or load main class %5$s%6$s%n
Caused by: java.lang.ClassNotFoundException: %5$s%6$s%n

```

--- a/itests/helloworld.java
+++ b/itests/helloworld.java
@@ -1,6 +1,6 @@
 ///usr/bin/env jbang "$0" "$@" ; exit $?
 // //DEPS <dependency1> <dependency2>
-//JAVA_OPTIONS -Dfoo=bar -Dbar=foo
+//JAVA_OPTIONS -Dfoo=bar "-Dbar=aap noot mies"
 
 import static java.lang.System.*;
 

--- a/src/main/java/dev/jbang/cli/BaseBuildCommand.java
+++ b/src/main/java/dev/jbang/cli/BaseBuildCommand.java
@@ -17,6 +17,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -332,7 +333,7 @@ public abstract class BaseBuildCommand extends BaseScriptDepsCommand {
 		// When persistent JVM args are set they override any runtime options set on the
 		// Source
 		List<String> rtArgs = ctx.getRuntimeOptionsOr(src);
-		String runtimeOpts = String.join(" ", rtArgs);
+		String runtimeOpts = String.join(" ", escapeArguments(rtArgs));
 		if (!runtimeOpts.isEmpty()) {
 			manifest.getMainAttributes()
 					.putValue(Source.ATTR_JBANG_JAVA_OPTIONS, runtimeOpts);
@@ -432,6 +433,62 @@ public abstract class BaseBuildCommand extends BaseScriptDepsCommand {
 		} else {
 			return cmd;
 		}
+	}
+
+	// NB: This might not be a definitive list of safe characters
+	static Pattern cmdSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
+	// TODO: Figure out what the real list of safe characters is for PowerShell
+	static Pattern pwrSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
+	static Pattern shellSafeChars = Pattern.compile("[a-zA-Z0-9._+=:@%/-]*");
+
+	/**
+	 * Escapes list of arguments where necessary using the current OS' way of
+	 * escaping
+	 */
+	static List<String> escapeOSArguments(List<String> args) {
+		return args.stream().map(BaseBuildCommand::escapeOSArgument).collect(Collectors.toList());
+	}
+
+	/**
+	 * Escapes list of arguments where necessary using a generic way of escaping
+	 * (we'll just be using the Unix way)
+	 */
+	static List<String> escapeArguments(List<String> args) {
+		return args.stream().map(BaseBuildCommand::escapeUnixArgument).collect(Collectors.toList());
+	}
+
+	static String escapeOSArgument(String arg) {
+		if (Util.isWindows()) {
+			arg = escapeWindowsArgument(arg);
+		} else {
+			arg = escapeUnixArgument(arg);
+		}
+		return arg;
+	}
+
+	static String escapeUnixArgument(String arg) {
+		if (!shellSafeChars.matcher(arg).matches()) {
+			arg = arg.replaceAll("(['])", "'\\\\''");
+			arg = "'" + arg + "'";
+		}
+		return arg;
+	}
+
+	static String escapeWindowsArgument(String arg) {
+		if (Util.isUsingPowerShell()) {
+			if (!pwrSafeChars.matcher(arg).matches()) {
+				arg = arg.replaceAll("(['])", "''");
+				arg = "'" + arg + "'";
+			}
+		} else {
+			if (!cmdSafeChars.matcher(arg).matches()) {
+				// Windows quoting is just weird
+				arg = arg.replaceAll("([()!^<>&|% ])", "^$1");
+				arg = arg.replaceAll("([\"])", "\\\\^$1");
+				arg = "^\"" + arg + "^\"";
+			}
+		}
+		return arg;
 	}
 
 }

--- a/src/main/java/dev/jbang/cli/Edit.java
+++ b/src/main/java/dev/jbang/cli/Edit.java
@@ -1,7 +1,7 @@
 package dev.jbang.cli;
 
 import static dev.jbang.Settings.CP_SEPARATOR;
-import static dev.jbang.cli.Run.escapeArguments;
+import static dev.jbang.cli.BaseBuildCommand.escapeOSArguments;
 import static dev.jbang.util.Util.isWindows;
 import static dev.jbang.util.Util.verboseMsg;
 import static java.lang.System.out;
@@ -82,7 +82,7 @@ public class Edit extends BaseScriptDepsCommand {
 				optionList.add(project.getAbsolutePath());
 
 				String[] cmd;
-				final String editorCommand = escapeArguments(optionList).stream().collect(Collectors.joining(" "));
+				final String editorCommand = escapeOSArguments(optionList).stream().collect(Collectors.joining(" "));
 				if (isWindows()) {
 					cmd = new String[] { "cmd", "/c", editorCommand };
 				} else {

--- a/src/main/java/dev/jbang/cli/Run.java
+++ b/src/main/java/dev/jbang/cli/Run.java
@@ -110,7 +110,7 @@ public class Run extends BaseBuildCommand {
 
 	String generateOSCommandLine(Source src, RunContext ctx) throws IOException {
 		List<String> fullArgs = generateCommandLineList(src, ctx);
-		String args = String.join(" ", escapeArguments(fullArgs));
+		String args = String.join(" ", escapeOSArguments(fullArgs));
 		// This avoids long classpath problem on Windows.
 		// @file is only available from java 9 onwards.
 		if (args.length() > COMMAND_LINE_LENGTH_LIMIT && Util.isWindows()
@@ -134,7 +134,7 @@ public class Run extends BaseBuildCommand {
 
 	String generateCommandLine(Source src, RunContext ctx) throws IOException {
 		List<String> fullArgs = generateCommandLineList(src, ctx);
-		return String.join(" ", escapeArguments(fullArgs));
+		return String.join(" ", escapeOSArguments(fullArgs));
 	}
 
 	List<String> generateCommandLineList(Source src, RunContext ctx) throws IOException {
@@ -330,42 +330,6 @@ public class Run extends BaseBuildCommand {
 
 	private void addJavaArgs(List<String> args, List<String> result) {
 		result.addAll(args);
-	}
-
-	static List<String> escapeArguments(List<String> args) {
-		return args.stream().map(Run::escapeArgument).collect(Collectors.toList());
-	}
-
-	// NB: This might not be a definitive list of safe characters
-	static Pattern cmdSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
-
-	// TODO: Figure out what the real list of safe characters is for PowerShell
-	static Pattern pwrSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
-
-	static Pattern shellSafeChars = Pattern.compile("[a-zA-Z0-9._+=:@%/-]*");
-
-	static String escapeArgument(String arg) {
-		if (Util.isWindows()) {
-			if (Util.isUsingPowerShell()) {
-				if (!pwrSafeChars.matcher(arg).matches()) {
-					arg = arg.replaceAll("(['])", "''");
-					return "'" + arg + "'";
-				}
-			} else {
-				if (!cmdSafeChars.matcher(arg).matches()) {
-					// Windows quoting is just weird
-					arg = arg.replaceAll("([()!^<>&|% ])", "^$1");
-					arg = arg.replaceAll("([\"])", "\\\\^$1");
-					return "^\"" + arg + "^\"";
-				}
-			}
-		} else {
-			if (!shellSafeChars.matcher(arg).matches()) {
-				arg = arg.replaceAll("(['])", "'\\\\''");
-				return "'" + arg + "'";
-			}
-		}
-		return arg;
 	}
 
 	static String generateMain(String main) {

--- a/src/test/java/dev/jbang/cli/TestRun.java
+++ b/src/test/java/dev/jbang/cli/TestRun.java
@@ -107,12 +107,15 @@ public class TestRun extends BaseTest {
 		assertThat(result, endsWith("helloworld"));
 		assertThat(result, containsString("classpath"));
 		assertThat(result, containsString(".jar"));
-		assertThat(result, containsString("-Dfoo=bar -Dbar=foo"));
+		assertThat(result, containsString("-Dfoo=bar"));
+		assertThat(result, containsString(BaseBuildCommand.escapeOSArgument("-Dbar=aap noot mies")));
 		// Make sure the opts only appear once
-		assertThat(result.replaceFirst(Pattern.quote("-Dfoo=bar -Dbar=foo"), ""),
-				not(containsString("-Dfoo=bar -Dbar=foo")));
+		assertThat(result.replaceFirst(Pattern.quote("-Dfoo=bar"), ""),
+				not(containsString("-Dfoo=bar")));
+		assertThat(result.replaceFirst(Pattern.quote("-Dbar=aap noot mies"), ""),
+				not(containsString("-Dbar=aap noot mies")));
 		// Make sure the opts only appear unquoted
-		assertThat(result, not(containsString(Run.escapeArgument("-Dfoo=bar -Dbar=foo"))));
+		assertThat(result, not(containsString(BaseBuildCommand.escapeOSArgument("-Dfoo=bar -Dbar=aap noot mies"))));
 		// assertThat(result, containsString("--source 11"));
 	}
 

